src/python/pants/backend/python/rules/pex_from_targets.py

```diff
@@ -121,6 +125,43 @@ async def pex_from_targets(request: PexFromTargetsRequest, python_setup: PythonS
         additional_requirements=request.additional_requirements,
     )
 
+    if python_setup.options.resolve_all_constraints:
+        if python_setup.requirement_constraints is None:
+            raise ValueError(
+                "resolve_all_constraints in the [python-setup] scope is set, so "
+                "requirement_constraints in [python-setup] must also be provided."
+            )
+        exact_req_projects = {Requirement.parse(req).project_name for req in exact_reqs}
+        # Add the requirements from the constraint file. Note that this only guarantees
+        # a single resolve if the constraint file covers all the individual requirements
+        # of individual targets. This may be defeated by, e.g., inline requirements.
+        # However we will still do the right thing in that case, it's just the performance
+        # will be impacted by extra resolves.
+        constraint_file_snapshot = await Get(
+            Snapshot,
+            PathGlobs(
+                [python_setup.requirement_constraints],
+                glob_match_error_behavior=GlobMatchErrorBehavior.error,
+                conjunction=GlobExpansionConjunction.all_match,
+                description_of_origin="the option `--python-setup-requirement-constraints`",
+            ),
+        )
+        constraints_file_contents = await Get(
+            DigestContents, Digest, constraint_file_snapshot.digest
+        )
+        constraints_file_reqs = set(
+            parse_requirements(next(iter(constraints_file_contents)).content.decode())
+        )
+        constraint_file_projects = {req.project_name for req in constraints_file_reqs}
+
+        if exact_req_projects.issubset(constraint_file_projects):
+            requirements = PexRequirements(str(req) for req in constraints_file_reqs)
+        else:
+            requirements = exact_reqs
```

Comment:
This would mean that the lockfile was incorrectly or partially created, I think? That probably justifies a warning.

Reply:
That could possibly the case. Or they may intentionally be using an inline requirement and they don't want it included in the global constraints file. I'd bias towards no warning (noisy), but will defer to you both.

Reply:
Given that 1) creating a lockfile is currently a manual process that someone can get wrong, and 2) the reason that you would do it would be to give you build determinism... I think that it makes sense to warn here, at the very least.

But perhaps it is the case that that warning should trigger regardless of the `--resolve-all-constraints` setting, anytime that you have a lockfile configured and we encounter a requirement that is not specified in the constraints file? This is in the general category of "things are pretty manual right now" as described in https://www.pantsbuild.org/docs/python-third-party-dependencies#using-a-lockfile-recommended

Reply:
Good points. That makes sense and fits with a theme I'd like for Pants to have of Pants eagerly helping you to do what you likely intended, kind of like how people describe the Rust compiler.

Reply:
Yeah, I think this makes sense as a warning. Will add.